### PR TITLE
Generic hash test plan

### PR DESCRIPTION
**Pull request by congh-nvidia**
### Description of PR

Summary:
This is the test plan for the feature generic hash.

### Type of change

- [ ] Bug fix
- [ ] Testbed and Framework(new/improvement)
- [x] Test case(new/improvement)

### Back port request
- [ ] 201911
- [ ] 202012
- [x] 202205

### CODE REVIEW

File: docs/testplan/Generic-hash-test-plan.md [hunk 1]
```diff
@@ -0,0 +1,314 @@
...
+
+### 2.1 The Generic Hash feature supports the following functionality:
+1. Ethernet packet hashing configuration with inner/outer IP frames
+2. Global switch hash configuration for ECMP and LAG
+3. Warm/Fast reboot
+
+### 2.2 This feature will support the following commands:
+
+1. config: set switch hash global configuration
+2. show: display switch hash global configuration
```
Comment (nhe-NV) [inline docs/testplan/Generic-hash-test-plan.md]: can also show the switch hash capabilities, maybe it should also be described here

File: docs/testplan/Generic-hash-test-plan.md [hunk 2]
```diff
@@ -0,0 +1,314 @@
...
+
+Notes: 
+  1. The hash field is randomly selected in the test cases. Currently these fields are tested: 'IN_PORT', 'SRC_MAC', 'DST_MAC', 'ETHERTYPE', 'VLAN_ID', 'IP_PROTOCOL', 'SRC_IP', 'DST_IP', 'L4_SRC_PORT', 'L4_DST_PORT', 'INNER_SRC_IP', 'INNER_DST_IP'.
+  2. DST_MAC, ETHERTYPE, VLAN_ID fields are only tested in lag hash test cases, because L2 traffic is needed to test these fields, and there is no ecmp hash when switching in L2.
+  3. IPv4 and IPv6 are covered in the test, but the versions(including the inner version when testing the inner fields) are randomly selected in the test cases.
+  4. For the inner fields, three types of encapsulations are covered: IPinIP, VxLAN and NVGRE. For the VxLAN packet, the default port 4789 and a custom port 13330 are covered in the test.
+
+### Test cases #1 - Verify the “show switch-hash capabilities” gets the supported hash fields.
+1. Get the supported hash fields via cli "show switch-hash capabilities"
+2. Check the fields are as expected`.
```
Comment (nazariig) [inline docs/testplan/Generic-hash-test-plan.md:226]: @congh-nvidia how it is supposed to be vendor agnostic? A dictionary?

File: docs/testplan/Generic-hash-test-plan.md [hunk 3]
```diff
@@ -0,0 +1,314 @@
...
+3. Configure the lag hash list to exclude the selected field to verify the lag hash does not affect the hash result.
+4. Send traffic with changing values of the field under test from a downlink ptf port to uplink destination via multiple nexthops.
+5. Check the traffic is balanced over the nexthops.
+6. If the uplinks are portchannels with multiple members, check the traffic is not balanced over the members.
+
+### Test cases #3 - Verify when generic lag hash is configured, the traffic can be balanced accordingly.
+1. The test is using the default links and routes in a t0/t1 testbed, and only runs on setups which have multi-member portchannel uplinks.
+2. Randomly select a hash field and configure it to the lag hash list via cli "config switch-hash global lag-hash".
+3. Configure the ecmp hash list to exclude the selected field to verify the ecmp hash does not affect the hash result.
+4. If the hash field is DST_MAC, ETHERTYPE or VLAN_ID, change the topology to allow L2 switching and send L2 traffic in the next step.
```
Comment (nazariig) [inline docs/testplan/Generic-hash-test-plan.md:240]: @congh-nvidia what does it mean "change topology"? IMHO, it doesn't efficient during test execution...

File: docs/testplan/Generic-hash-test-plan.md [hunk 4]
```diff
@@ -0,0 +1,314 @@
...
+6. Check the traffic is balance over all the uplink ports.
+7. Randomly do fast/warm/cold reboot.
+8. After the reboot, check the generic hash config via cli, it should keep the same as before the reboot.
+9. Send traffic again.
+10. Check the traffic is balance over all the uplink ports.
+
+### Test cases #8 - Verify the generic hash cannot be configured successfully with invalid parameters.
+1. Configure the ecmp/lag hash with invalid fields list parameter.
+2. Check there is a cli error that notifies the user the parameter is invalid.
+3. Check there is a warning printed in the syslog.
```
Comment (nazariig) [inline docs/testplan/Generic-hash-test-plan.md]: @congh-nvidia FYI. This won't happen each time the CLI input is invalid, but only at some specific circumstances, e.g. config was valid but it failed to get to the DB somehow.

File: docs/testplan/Generic-hash-test-plan.md [hunk 5]
```diff
@@ -0,0 +1,314 @@
...
+4. Check the running config is not changed.
+5. The invalid parameters to test:
+  a. empty parameter
+  b. single invalid field
+  c. invalid fields combined with valid fields
+  d. duplicated valid fields(depends on the final implememtation)
+
+### Test cases #9 - Verify when a generic hash config key is removed, or updated with invalid values from config_DB via redis cli, there will be warnings printed in the log.
```
Comment (nazariig) [inline docs/testplan/Generic-hash-test-plan.md:300]: @congh-nvidia do we really need to test this? The user interface is CLI only.

**Review by nazariig — CHANGES_REQUESTED** [review]

File: docs/testplan/Generic-hash-test-plan.md (hunk 2 above)
Comment (congh-nvidia) [reply docs/testplan/Generic-hash-test-plan.md:226]: Yes, by a dictionary with the vendor names as the key:
HASH_CAPABILITIES = {'mellanox': {'ecmp': MELLANOX_ECMP_HASH_FIELDS,
                                  'lag': MELLANOX_LAG_HASH_FIELDS},
                     'default': {'ecmp': DEFAULT_ECMP_HASH_FIELDS,
                                 'lag': DEFAULT_LAG_HASH_FIELDS}}
Other vendors need to add their own capabilities.

File: docs/testplan/Generic-hash-test-plan.md (hunk 3 above)
Comment (congh-nvidia) [reply docs/testplan/Generic-hash-test-plan.md:240]: The community testbeds only have L3 configurations, if we want to test L2 traffic, need to do configure some L2 ports and vlan, it doesn't cost much time. No need to re-deploy sonic topology.

File: docs/testplan/Generic-hash-test-plan.md (hunk 4 above)
Comment (congh-nvidia) [reply docs/testplan/Generic-hash-test-plan.md]: You're right, the syslog check should be removed in this test case.

File: docs/testplan/Generic-hash-test-plan.md (hunk 2 above)
Comment (nazariig) [reply docs/testplan/Generic-hash-test-plan.md:226]: @congh-nvidia ok

File: docs/testplan/Generic-hash-test-plan.md (hunk 3 above)
Comment (nazariig) [reply docs/testplan/Generic-hash-test-plan.md:240]: @congh-nvidia is it reasonable to have a separate test case for that? Just to not mix a lot of stuff for the same scenario.

File: docs/testplan/Generic-hash-test-plan.md (hunk 4 above)
Comment (nazariig) [reply docs/testplan/Generic-hash-test-plan.md]: @congh-nvidia ok

File: docs/testplan/Generic-hash-test-plan.md (hunk 5 above)
Comment (congh-nvidia) [reply docs/testplan/Generic-hash-test-plan.md:300]: This is required by the QG.

Comment (nazariig) [reply docs/testplan/Generic-hash-test-plan.md:300]: @congh-nvidia got it.

File: docs/testplan/Generic-hash-test-plan.md (hunk 3 above)
Comment (congh-nvidia) [reply docs/testplan/Generic-hash-test-plan.md:240]: To save the execution time, we are not aiming to cover all the possible scenarios and cover all fields in a single run, the hash field is generated randomly. That's why all the fields are tested in a single test case.

File: docs/testplan/Generic-hash-test-plan.md (hunk 1 above)
Comment (congh-nvidia) [reply docs/testplan/Generic-hash-test-plan.md]: Done

**Review by nazariig — APPROVED** [review]

**Pull request closed without merging**